### PR TITLE
[AIR] Add mosaic trainer skeleton

### PR DESCRIPTION
**Pull request by ilee300a**
Signed-off-by: ilee300a <ilee300@anyscale.com>

In this PR, we provide initial commits for integrating Mosaic library with Ray. As Mosaic library provides algorithmic acceleration, providing further acceleration from the system side via Ray's distributed training, we can improve the speedup of training process.

Included in this PR is `MosaicTrainer` skeleton code, which currently can run a basic composer trainer without any logging.

### `MosaicTrainer` design
`MosaicTrainer` provides a wrapper for creating a worker loop that creates `composer.trainer.Trainer` and runs it.

With ``MosaicTrainer``, there will always be checkpointing, with a default ``CheckpointSaver`` that saves checkpoints to `ray_tmp` folder. All ``CheckpointSaver`` objects are wrapped with ``RayTrainReportCallback`` to allow reporting the paths of saved checkpoints. (For more detail check out ``RayTrainReportCallback``). When resuming training from checkpoint, if ``load_path`` is provided in ``trainer_init_config``, it will be used to load saved checkpoints. Otherwise, if ``resume_from_checkpoint`` is provided, the last reported path will be passed in as ``load_path`` to the Composer trainer.

All logged information, whether there is a composer logger provided or not, is reported as metrics via ``RayLogger``. ``RayLogger`` is added to the logger configuration for Composer trainer initialization. (For more detail check out ``RayLogger``.)

#### `_mosaic_train_loop_per_worker`
Pre-worker training loop for Mosaic Composers
- sets environment variables for DDP
- gets dataset shards and converts them to an Iterable that can later be converted into `composer.core.DataSpec`
- create `composer.trainer.Trainer` and runs it

#### `_mosaic_utils.process_datasets`
Process ray dataset shard into an iterable that can be converted into `composer.core.DataSpec`
- requires `batch_size : int` and `labels : List[str]` to be included in `trainer_init_config`
- Implementation
    - `RayDatasetMosaicIterable`
        - a wrapper class that provides iterator over ray dataset
    - `_mosaic_iterator`
        - iterator class for composer trainer that iterates over ray dataset.
        - uses `iter_torch_batches` to get each batch
        - requires that the dataset is mapped into pandas data frame with appropriate labels.
        - rearranges the DataFrame data into a list of each column data in the order the column labels are provided.
            - Enables the user to customize the data included in each batch.
            - Provides flexibility as the type and the number of data is clearly defined by the user.
            - compatible with composer’s training loop as the format of mini batch extraction training is `input, _ = batch` and `_, target = batch`

#### `RayLogger`
A logger to relay all logged information to ray. This logger allows utilizing all necessary logging and logged data handling provided by ray. When some information is logged, then the backing data dictionary is updated. The most up-to-date data dictionary is reported as metrics every `batch_checkpoint` and `epoch_checkpoint`

#### `RayTrainReportCallback`
A callback that wraps Composer's ``CheckpointSaver``. This class is used to wrap each Composer ``CheckpointSaver`` to be used in Composer trainer. The main role of this callback is to report the paths of the checkpoints saved by the Composer ``CheckpointSaver`` it wraps. In addition, when the training ends, (either with or without exception) the last checkpoint and the list of all the checkpoints that have been saved and the list of Composer loggers are reported.

### Initial Experiments and Results
Experiment Setting(Omit all the hyperparameters unrelated to training speed):
- machine : g4dn.4xlarge
- num_workers : 4
- Dataset : CIFAR 10
- Model : “resnet_56” from composer library
- Batch size : 64
- Num Epochs : 3

Experiment Result
- measured end-to-end training time (no eval phase)
- script example
    ```jsx
    start_time = time.perf_counter()
    trainer.fit()
    end_time = time.perf_counter()
    print(f"Total training time : {end_time - start_time:0.4f} seconds")
    ```

Trainer type | Total Time (s)
-- | --
Mosaic Composer | 153.7415
Mosaic Composer + Ray | 95.9473

### Future Steps
- Added unit testing for MosaicTrainer but one for MosaicPredictor is still WIP.
- More tests will be added to the MosaicTrainer unit test

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/mosaic/_mosaic_utils.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+class RayDatasetMosaicIterable:
+    def __init__(self, dataset, batch_size, labels):
+        self.dataset = dataset
+        self.batch_size = batch_size
+        self.labels = labels
+        self.total_samples = dataset.count()
+
+    def __len__(self):
+        return math.ceil(self.dataset.count() / self.batch_size)
```
Comment (Yard1) [inline python/ray/train/mosaic/_mosaic_utils.py:36]: Use `total_samples` instead of calling count again?

Comment (ilee300a) [reply python/ray/train/mosaic/_mosaic_utils.py]: fixed!

File: python/ray/train/mosaic/_mosaic_utils.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        self.labels = labels
+        self.total_samples = dataset.count()
+        self.batch_iter = self.dataset.iter_torch_batches(batch_size=batch_size)
+
+    def __next__(self):
+        next_data = next(self.batch_iter)
+        return [next_data[label] for label in self.labels]
+
+
+class RayDatasetMosaicIterable:
```
Comment (xwjiang2010) [inline python/ray/train/mosaic/_mosaic_utils.py:27]: internal class only?
Also add some documentation?

File: python/ray/train/mosaic/_mosaic_utils.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            )
+        else:
+            super(RayTrainReportCallback, self).__init__(**args)
+
+    def close(self, state: State, logger: Logger) -> None:
+        del logger  # unused
+        checkpoint = Checkpoint.from_dict(
+            {
+                "last_checkpoint": self.last_checkpoint,
+                "loggers": self.loggers,
```
Comment (xwjiang2010) [inline python/ray/train/mosaic/_mosaic_utils.py:140]: hmmm, why reporting this?

Comment (xwjiang2010) [reply python/ray/train/mosaic/_mosaic_utils.py:140]: ah I see InMemoryLogger is used down below.
Should we only report with InMemoryLogger then?
In other words, is it guaranteed that the other loggers are serializable?

Comment (ilee300a) [reply python/ray/train/mosaic/_mosaic_utils.py]: It seems to be unnecessary to include loggers other than InMemoryLogger, so now we only report InMemoryLoggers

**Comment by Yard1** [comment]
Hey, is this ready for review? @ilee300a

**Comment by ilee300a** [comment]
> Hey, is this ready for review? @ilee300a
Currently I am adding more documentation and tests, but the main code changes should be good for review.

File: python/ray/train/mosaic/mosaic_predictor.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        cls,
+        path: Path,
+        model: torch.nn.Module,
+        strict: bool = False,
+        **predictor_kwargs
+    ):
+        """
+        This function creates a BatchPredictor from a saved Composer checkpoint.
+        Because Composer library is built on PyTorch, `TorchPredictor` is used
+        as the `predictor_cls`
```
Comment (Yard1) [inline python/ray/train/mosaic/mosaic_predictor.py:30]:
```suggestion
        Because Composer library is built on PyTorch, ``TorchPredictor`` is used
        as the ``predictor_cls``
```

File: python/ray/train/mosaic/mosaic_predictor.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        Args:
+            path: Path to the saved checkpoint object
+            model: the model to which the saved checkpoint will be loaded. This
+                model should be a bare torch model, without any composer wrapping.
+            strict: the boolean variable for strict state_dict loading onto the model
+        """
+        model_state_dict = torch.load(path)["state"]["model"]
+
+        # remove module prefixes when loading the model weights
+        while True:
```
Comment (Yard1) [inline python/ray/train/mosaic/mosaic_predictor.py:41]: isn't there a built-in torch util for this? I may be misremembering though

File: python/ray/train/mosaic/mosaic_predictor.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
+from ray.train.torch import TorchPredictor
+from ray.train.batch_predictor import BatchPredictor
+from ray.air.checkpoint import Checkpoint
+from pathlib import Path
+import torch
+
+
+class MosaicPredictor(BatchPredictor):
```
Comment (Yard1) [inline python/ray/train/mosaic/mosaic_predictor.py:8]: Why are we wrapping around BatchPredictor here? This is not a pattern we have anywhere else. Can we subclass TorchPredictor instead?

File: python/ray/train/mosaic/_mosaic_utils.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+                overwrite=checkpoint_saver.overwrite,
+                num_checkpoints_to_keep=checkpoint_saver.num_checkpoints_to_keep,
+                weights_only=checkpoint_saver.weights_only,
+                **args
+            )
+        else:
+            super(RayTrainReportCallback, self).__init__(**args)
+
+    def close(self, state: State, logger: Logger) -> None:
+        del logger  # unused
```
Comment (Yard1) [inline python/ray/train/mosaic/_mosaic_utils.py:195]: is this necessary? I don't think we should explicitly delete variables

File: python/ray/train/mosaic/_mosaic_utils.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+class RayTrainReportCallback(CheckpointSaver):
+    """A callback that wraps Composer's ``CheckpointSaver``.
+
+    This class is used to wrap each Composer ``CheckpointSaver`` to be used in Composer
+    trainer. The main role of this callback is to report the paths of the checkpoints
+    saved by the Composer ``CheckpointSaver`` it wraps. In addition, when the training
+    ends, (either with or without exception) the last checkpoint and the list of all
+    the checkpoints that have been saved and the list of Composer InMemoryLogger are
+    reported.
```
Comment (Yard1) [inline python/ray/train/mosaic/_mosaic_utils.py:142]: Shouldn't we instead create checkpoints from those paths (`Checkpoint.from_directory`) and report those? I am not sure how they are going to be transferred to other nodes in the current logic. It seems like it will only work on a single node to me.

File: python/ray/train/mosaic/_mosaic_utils.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+    def _update_checkpoint(self, state: State):
+        # check that the saved checkpoint is not redundant
+        if len(self.saved_checkpoints) > self.checkpoint_count:
+            self.last_checkpoint = [
+                chkpt_path.absolute() for chkpt_path in self.saved_checkpoints[-1][1]
+            ]
+            self.checkpoint_count = len(self.saved_checkpoints)
+            session.report(
+                metrics={},
```
Comment (Yard1) [inline python/ray/train/mosaic/_mosaic_utils.py:221]: We shouldn't report checkpoints separately from metrics - we want them to be associated with those so that we can determine the best checkpoint to keep. We also would like to keep parity between training iterations and the number of reports made. Can we somehow combine metrics and checkpoint reporting here?

**Review by Yard1 — COMMENTED** [review]
Left some comments. My main question is about checkpointing & reporting.

File: python/ray/train/mosaic/mosaic_predictor.py (hunk 5 above)
Comment (ilee300a) [reply python/ray/train/mosaic/mosaic_predictor.py:41]: There is but I purposefully did not use it because that function has extra things that is not needed for our case. (Intended to reduce the overhead of having to call another function + spending time doing unnecessary tasks)

File: python/ray/train/mosaic/mosaic_predictor.py (hunk 6 above)
Comment (ilee300a) [reply python/ray/train/mosaic/mosaic_predictor.py:8]: I wrapped around BatchPredictor considering the usability. In most cases prediction would be made via BatchPredictor, and even when we wrap around TorchPredictor, it would not provide any more useful information. The main problem with getting a predictor for Mosaic Composer is that the checkpoint has to be reformatted, and so by directly wrapping around BatchPredictor, we can more efficiently reformat the checkpoint and instantiate BatchPredictor right away

File: python/ray/train/mosaic/_mosaic_utils.py (hunk 7 above)
Comment (ilee300a) [reply python/ray/train/mosaic/_mosaic_utils.py:195]: This is the general practice shown in the Composer library, so I followed their practice. This can be removed if deemed extra

File: python/ray/train/mosaic/mosaic_predictor.py (hunk 5 above)
Comment (Yard1) [reply python/ray/train/mosaic/mosaic_predictor.py:41]: Is there actually a noticeable performance difference? If not, I'd reuse existing code. We don't need to shave every millisecond :)

File: python/ray/train/mosaic/mosaic_predictor.py (hunk 6 above)
Comment (Yard1) [reply python/ray/train/mosaic/mosaic_predictor.py:8]: I am not sure that's the pattern we want (every other framework passes their predictor to batch predictor), but happy to discuss this offline

File: python/ray/train/mosaic/_mosaic_utils.py (hunk 7 above)
Comment (Yard1) [reply python/ray/train/mosaic/_mosaic_utils.py:195]: If this is the practice they are using, then that's fine with me, though I'd add a comment somewhere (could be just at the top of the file/before the class) that this is their practice.

File: python/ray/train/mosaic/mosaic_predictor.py (hunk 5 above)
Comment (ilee300a) [reply python/ray/train/mosaic/mosaic_predictor.py:41]: I have not really measured the performance. However, one advantage of using this code instead of the existing function is that we can easily note whether a prefix has been removed or not. The existing function does not return anything, so we would have to iterate through and see if there are any changes from the previous version.

Comment (Yard1) [reply python/ray/train/mosaic/mosaic_predictor.py:41]: That's fair! let's make sure to add a short comment explaining why we aren't using the built-in function then

File: python/ray/train/mosaic/mosaic_predictor.py (hunk 6 above)
Comment (ilee300a) [reply python/ray/train/mosaic/mosaic_predictor.py]: MosaicPredictor has been updated
